Hold data for all relations in a single InMemoryLayer.

hlinnaka opened the pull request:
This is a small step towards turning the storage format into an LSM tree.
In an LSM tree, the MemTable - or InMemoryLayer in our terms - holds data
for the whole key range.  Checkpoint still writes it out to multiple
per-segment layers, however, so this commit the on-disk format yet.

Move code to handle rel size updates from InMemoryLayer to the caller.
When you modify a block of a relish or segment for the first time,
the InMemoryLayer doesn't know its old size. The caller needs to find it
out and tell it.

Merge InMemoryLayer::put_wal_record and put_page_image into a single
put_page_version function.

hlinnaka commented:
As the commit message says, one change here is the way we track relation size. Previously, if you wrote to a block beyond EOF, the code in InMemoryLayer filled the "gap" with zeros. For example, if a relation had size 10, and you wrote to block 12, block 11 was filled with zeros. However, because that was done in InMemoryLayer, which used to deal with a single segment, that didn't work correctly at a segment boundary. In this PR, I moved that gap-filling logic, and tracking the relation size in general, to LayeredRepository. That's better, but I'm still not 100% happy with it. It still seems complicated. Maybe the relation size tracking should be lifted further up in the stack, from layered_repository.rs to walingest.rs.

Another thing I'm not 100% happy with is how we deal with "non-blocky" relations. Before this PR, a non-blocky relation always implicitly had size of 1 block, and that block could be any size. That's still true with this PR, but for simplicity, the same size-tracking logic is now applied to non-blocky relations too, even though it's a bit pointless. So we now store the size, which is always 1 block, in the in-memory and delta layers of non-blocky relations too. I've got a feeling that things would be simpler, if we assumed that all blocks are always 8kB in size, and chopped non-blocky relishes into 8kB blocks too. That would add a layer of code, maybe in walingest.rs, to chop non-blocky relishes into 8kB blocks and to reassemble them again when needed, but that way, the repository code in layered_repository.rs could treat all relishes the same. Anyway, that's not part of this PR, but a refactoring we might want to do separately in the future.

bojanserafimov commented on pageserver/src/layered_repository/page_versions.rs:
```diff
@@ -1,268 +0,0 @@
-}
-
-#[cfg(test)]
-mod tests {
```
are we gonna miss these unit tests? Are they copied over somewhere?

hlinnaka replied:
No, the test was testing the 'page_versions::ordered_page_version_iter' function, but the function is gone now.

bojanserafimov reviewed (commented):
Too much stuff going on. Maybe there's a more tactical way to make this change that's easier to review. Anyway, I didn't get to the important bits yet I but left you some comments to give you an idea of what trips me up when I read.

hlinnaka commented on pageserver/src/repository.rs:
```diff
@@ -746,8 +743,8 @@ mod tests {
         for blknum in 0..pg_constants::RELSEG_SIZE + 1 {
-            let img = TEST_IMG(&format!("foo blk {} at {}", blknum, Lsn(lsn)));
             lsn += 0x10;
+            let img = TEST_IMG(&format!("foo blk {} at {}", blknum, Lsn(lsn)));
```
The LSN in "foo blk {} at {}" string that is stored in the content of the page versions was wrong, because we advanced 'lsn' after creating the string. That was unrelated to this PR really, I just happened to spot it while working on this.

I opened a separate PR for that: https://github.com/zenithdb/zenith/pull/1315

hlinnaka commented on pageserver/src/layered_repository/inmemory_layer.rs:
```diff
@@ -1,37 +1,51 @@
+//! modifications to the layer in the LSN range. If a segment is not covered by the
+//! layer, there has been no modifications to it in the layer's LSN range, and you
+//! should check the predecessor layer. You can check if a segment is covered by
+//! calling 'covers_seg', which
```
Fixed

hlinnaka commented on pageserver/src/repository.rs:
```diff
@@ -1171,6 +1168,7 @@ mod tests {
         // Create a timeline with disk_consistent_lsn = 8000
         let tline = repo.create_empty_timeline(TIMELINE_ID, Lsn(0x8000))?;
         let writer = tline.writer();
+        writer.put_page_image(TESTREL_A, 0, Lsn(0x8000), TEST_IMG("foo blk 0 at 0x8000"))?;
```
I added a comment. I think this actually to be fixed before pushing, but at least there's a FIXME comment explaining it now.

hlinnaka commented on pageserver/src/layered_repository/inmemory_layer.rs:
```diff
@@ -308,23 +400,27 @@ impl Layer for InMemoryLayer {
-        }
+                    println!("blk {} at {}: {}\n", blknum, lsn, pv_description);
+                }
+        */
```
I broke the dump() function and just commented it out instead of fixing it. It's just for debugging purposes and isn't actually used anywhere. The dump() functions for Delta and Imagelayers are used by the 'dump_layerfile' utility, but the InMemoryLayer::dump function is completely unused. It can come handy while debugging though. In any case, fixed it now.

hlinnaka commented on pageserver/src/layered_repository/layer_map.rs:
```diff
@@ -38,16 +32,11 @@ lazy_static! {
-    current_generation: u64,
+    pub next_open_layer_at: Option<Lsn>,
+    pub open_layer: Option<Arc<InMemoryLayer>>,
+    pub frozen_layer: Option<Arc<InMemoryLayer>>,
```
Added comments

hlinnaka commented on pageserver/src/layered_repository/layer_map.rs:
```diff
@@ -148,42 +81,63 @@ impl LayerMap {
         NUM_ONDISK_LAYERS.dec();
     }
 
+    fn filter_fn(query: Option<RelTag>, seg: SegmentTag) -> bool {
```
Refactored

hlinnaka commented on pageserver/src/layered_repository/inmemory_layer.rs:
```diff
@@ -57,58 +71,133 @@ pub struct InMemoryLayer {
+}
+
+#[derive(Default)]
+struct PerSeg {
```
I added some comments for now. I'm all ears for a better name.

hlinnaka commented on pageserver/src/layered_repository/inmemory_layer.rs:
```diff
@@ -57,58 +71,133 @@ pub struct InMemoryLayer {
+    Create(SegmentBlk),
+    Size(SegmentBlk),
+    Drop,
+}
```
Added

hlinnaka commented:
> Too much stuff going on. Maybe there's a more tactical way to make this change that's easier to review. Anyway, I didn't get to the important bits yet I but left you some comments to give you an idea of what trips me up when I read.

Thanks! I addressed the comments you left so far.

Yeah, this is a large patch :-(. I'll think about how this could be split up into smaller pieces. Maybe the removal of 'page_versions.rs' and the refactoring related to that could be extracted to a separate PR, I'll try that.

hlinnaka commented:
Rebased. And I squashed the fixes to the 'InMemoryLayer::dump'  function to the main commit.

hlinnaka commented:
This has been superseded by https://github.com/zenithdb/zenith/pull/1350
